### PR TITLE
Added utility class for calculating basic statistical metrics

### LINKED ISSUE #1: Implement basic statistic metrics
- [ ] <b>Description</b>
    Implement basic statistic metrics:
    * Mean
        * Arithmetic mean
        * Geometric mean
        * Harmonic mean
   * Median
   * Mode
   * Standard deviation
   * Variance
   * Frequency distribution (map with the input collection values as keys and the count/frequency as values)
        *  Top n frequent elements
   * Minimum/max value
        * Top n maximum/minimum elements
   It should be able to accept a collection/array as input.
- [ ] <b>Use cases</b>
    Data understanding
- [ ] <b>Proposed design</b>
    Use parallel streams as much as possible.
- [ ] <b>Alternative design approaches</b>
    Use non parallel approaches.
- [ ] <b>Workaround</b>
   "Use a calculator" - Silvia
   "Or even better do it manually" - Ranjan
- [ ] <b>Additional information</b>
   N/A

### PR DESCRIPTION
**Pull request by silvs110**
## Description

Added StatisticsUtility to retrieve basic statistical metrics and its respective unit tests.

Closes #1.

## Author's Checklist
- [x] **These changes don't break existing dependants.** If they do, and you suggest to proceed, 
please explain why.

## Reviewer's Checklist
- [x] **All classes have documentation comments.**
- [x] **All methods have documentation comments.**
- [x] **No parameters are hardcoded.** Should be loaded as properties instead. If there needs to be hardcoded properties, explain why.
- [x] **All information to be logged/displayed to the user are internationalized.** If not, explain why. 
- [x] **README.md has been updated if needed.**
- [x] **pom.xml version is set to the latest date** If not, explain why.

## Assignee's Checklist
- [x] **All GitHub action checks have passed.**
- [x] **All reviewers have approved.**
- [x] **Relevant documentation has been updated if needed.**
- [x] **pom.xml version is set to the latest date** If not, explain why.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -3,11 +3,11 @@
...
   <modelVersion>4.0.0</modelVersion>
- 
+
   <!-- Update the following three entries -->
-  <groupId>org.padaiyal</groupId>
-  <artifactId>mavenprojecttemplate</artifactId>
-  <version>2021.04.26</version>
+  <groupId>org.padaiyal.utilities</groupId>
+  <artifactId>aayvalar</artifactId>
+  <version>2021.05.01</version>
```
Comment (the-c0d3br34k3r) [inline pom.xml]:
```suggestion
  <version>2021.05.02</version>
```

File: pom.xml [hunk 2]
```diff
@@ -3,11 +3,11 @@
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
   <modelVersion>4.0.0</modelVersion>
- 
+
   <!-- Update the following three entries -->
```
Comment (the-c0d3br34k3r) [inline pom.xml]:
```suggestion
```

File: src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.logging.log4j.Logger;
+import org.junit.jupiter.params.converter.ArgumentConversionException;
+import org.junit.jupiter.params.converter.SimpleArgumentConverter;
+import org.padaiyal.utilities.I18nUtility;
+import org.padaiyal.utilities.PropertyUtility;
+
+/**
+ * Converts a string into arrays of Strings.
+ */
+public class StringArrayConverter extends SimpleArgumentConverter {
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java]: We should include this with padaiyal/jUnitTestExtras.

File: src/main/java/org/padaiyal/aayvalar/statistics/MeanType.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package org.padaiyal.aayvalar.statistics;
+
+/**
+ * Enum used for representing the different types of mean.
+ */
+public enum MeanType {
+  ARITHMETIC,
+  GEOMETRIC,
+  HARMONIC,
+  DONOTUSE // for testing only https://stackoverflow.com/questions/46937118/force-switch-default-in-java-enum
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/statistics/MeanType.java]:
```suggestion
  DONOTUSE // for code coverage only https://stackoverflow.com/questions/46937118/force-switch-default-in-java-enum
```

File: src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      );
+    }
+  }
+
+  /**
+   * Converts a string into an array of specified types.
+   *
+   * @param source     The string to convert.
+   * @param targetType The type of array to convert it to.
+   * @throws ArgumentConversionException When the provided targetType is not supported.
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java]: Comments aren't aligned correctly.

File: src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  private StringArrayConverter() {
+
+  }
+
+  /**
+   * Converts a string into an array of specified types.
+   *
+   * @param arrayObject The string to convert.
+   * @param targetType  The type of array to convert it to.
+   * @throws ArgumentConversionException When the provided targetType is not supported.
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java]: Comments aren't aligned correctly.

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+    double collectionLength = inputCollection.size();
+    return switch (meanType) {
+      case ARITHMETIC -> //noinspection OptionalGetWithoutIsPresent
+          inputCollection.parallelStream()
+              .mapToDouble(Number::doubleValue)
+              .average()
+              .getAsDouble();
+      case GEOMETRIC -> Math.pow(
+          inputCollection.stream()
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]:
```suggestion
          inputCollection.parallelStream()
```

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+              .mapToDouble(Number::doubleValue)
+              .reduce(1, (subtotal, element) -> subtotal * element),
+          1 / collectionLength);
+      case HARMONIC -> collectionLength / (
+          inputCollection.stream()
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]:
```suggestion
          inputCollection.parallelStream()
```

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+  public static double getMedian(Collection<? extends Number> inputCollection) {
+    Objects.requireNonNull(inputCollection);
+
+    int collectionLength = inputCollection.size();
+    //noinspection OptionalGetWithoutIsPresent
+    return inputCollection.parallelStream()
+        .mapToDouble(Number::doubleValue)
+        .sorted()
+        .skip((collectionLength - 1) / 2)
+        .limit(2 - collectionLength % 2)
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: Good one!

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+          I18nUtility.getFormattedString(
+              "StatisticsUtility.error.invalidMeanType",
+              meanType.toString()
+          )
+      );
+    };
+  }
+
+  /**
+   * Gets the median of the provided collection.
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: Add a few words explaining median.

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+  /**
+   * Gets the average of the provided collection.
+   *
+   * @param inputCollection The collection to get the mean from.
+   * @param meanType        The type of mean to calculate.
+   * @return The mean of the values inside a collection.
+   */
+  public static double getMean(Collection<? extends Number> inputCollection, MeanType meanType) {
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: Add a few words explaining the different types of means, or at least link them to their corresponding wikipedia page..

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+      Collection<? extends Number> inputCollection, int topNumbers) {
+    Objects.requireNonNull(inputCollection);
+    if (topNumbers < 1) {
+      throw new IllegalArgumentException(
+          I18nUtility.getString("StatisticsUtility.error.negativeTopNumberIsProvided")
+      );
+    }
+
+    return getFrequencyDistribution(inputCollection)
+        .entrySet().parallelStream()
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]:
```suggestion
        .entrySet()
        .parallelStream()
```

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+   * Gets the standard deviation from the provided collection.
+   *
+   * @param inputCollection The input collection to retrieve the standard deviation.
+   * @param meanType        The mean type to use for calculation.
+   * @return The standard deviation from the provided collection.
+   */
+  public static double getStandardDeviation(Collection<? extends Number> inputCollection,
+      MeanType meanType) {
+    double variance = getVariance(inputCollection, meanType);
+    return Math.sqrt(variance);
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: Merge into one
```suggestion
    return Math.sqrt(getVariance(inputCollection, meanType));
```

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Reviewed. Requested changes.

- [x] Add wikipedia links for metrics like mean, median, variance etc in the documentation comment of the method implementing it.
- [x] Ensure that documentation comments are aligned correctly.
- [x] Convert tests like testGetTopFrequentElementsWithInvalidInput() to parametrized tests by using the ExceptionClassConverter.
- [x] Open an issue in jUnitTestExtras for adding StringArrayConverter. - https://github.com/padaiyal/jUnitTestExtras/issues/4

File: src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java (hunk 3 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java]: sure. I'll remove this once this class is moved to padaiyal/jUnitTestExtras

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java (hunk 8 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: Using parallel stream will not return the correct value. Not really sure why.

Comment (the-c0d3br34k3r) [reply src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: In the stream, map it to (1/element) and then do a sum().

File: src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java (hunk 5 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/parameterconverters/StringArrayConverter.java]: I think that it only aligns the parameters

File: src/main/java/org/padaiyal/aayvalar/statistics/MeanType.java (hunk 4 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/statistics/MeanType.java]: done

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java (hunk 11 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: done

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java (hunk 8 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: done

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java (hunk 7 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: done

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java (hunk 10 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: done

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java (hunk 12 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: done

File: src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java (hunk 13 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/aayvalar/statistics/StatisticsUtility.java]: done

File: pom.xml (hunk 1 above)
Comment (silvs110) [reply pom.xml]: done

**Review by the-c0d3br34k3r — APPROVED** [review]

**Pull request merged**